### ISSUE TITLE
Plugin-based system to allow 3rd party mods

### ISSUE THREAD

**Issue by pinkestt** [cursey/kanan-new#82]
I've been using Kanan for a while and I'd love to start developing some mods for it. However, I've noticed from looking through the source that the current mods are either patches loaded from a JSON file (not very powerful) or are compiled in (not very extensible).

I'm wondering if you would be interested in a system where external mods can be loaded from DLLs rather than needing to be compiled in. This would allow people to develop mods without needing to have them merged into upstream, which could be better for some more specialist or larger mods, as well as reducing the amount of recompilation required during development.

Looking at `Mods.cpp`, this doesn't seem like it would be a particularly significant change. I would be happy to submit a pull request that adds code to automatically load each DLL in a directory at the end of `Mods::loadMods`.

Let me know if this is wanted and I'll get hacking! ^_^

**Comment by Rietty** [comment]
There is already a feature in the Loader that loads/injects dlls that are
listed in a specific text file. The one drawback is that you won't be able
to really use the Kanan stuff like that.

**Comment by pinkestt** [comment]
Exactly. That's the motivation here: to allow using Kanan APIs without needing to compile your mods into Kanan itself.

**Comment by Rietty** [comment]
I mean you're welcome to do if you'd like to? It does sound neat. How do
you plan on using the Kanan API with your external mods however?

**Comment by cursey** [comment]
I often see plugin systems like what your suggesting to be an example of over-engineering. Most users don't care about a plugin system they just want to download the mod and have it work. Most developers have no problem forking kanan and developing their own changes on top of it. 

Personally I want to encourage upstream contributions and having a plugin system runs counter to that idea. Private mods don't interest me. People developing private mods with kanan is fine but it's not something I want to go out of my way to encourage. If you're interested in developing mods for kanan you shouldn't have any issue with doing it publicly where everyone can benefit.

Specifically, you say you want to start developing mods for Kanan. What about kanan is currently stopping you from doing that? A plugin system wouldn't make it any easier to get started than just forking kanan as it is.

**Comment by cursey** [comment]
I'm not totally against the idea I just need a better understanding of what the point is.

**Comment by cursey** [comment]
I won't turn down a PR for a plugin system if it doesn't complicate things. How were you planning on implementing it?

**Comment by pinkestt** [comment]
Here's how it'd work in C++-style pseudocode:

```c++
// Inside Mods::loadMods, or a method called from there
for (string& filename : listDirectory(PLUGIN_DIRECTORY)) {
    auto modLib = loadDLL(filename);
    auto mod = getSymbolPointer(modLib, "_kananModExport");
    addMod(mod);
}
```

Inside a header that's exposed to plugins, there would be a macro for creating the `_kananLoadPlugin` function, which would look something like:

```c++
#define KANAN_MOD(modClass) std::unique_pointer<modClass> _kananModExport = std::make_unique<modClass>();
```

So in your plugin, you'd do something like this:

```c++
#include <Kanan.h>

class MyMod : public Mod {
    // ...
}

KANAN_MOD(MyMod)
```

**Comment by pinkestt** [comment]
If all goes well, it shouldn't require any changes other than adding a few lines in `Mods::loadMods` and possibly moving some header files around to make them easier to access from plugins.

**Comment by pinkestt** [comment]
I had a play around with loading DLLs using `LoadLibrary` and `GetProcAddress` from the Windows API, and it seems to work well. I'll try integrating it with Kanan later on.

**Comment by LoneDev6** [comment]
> 
> 
> I've been using Kanan for a while and I'd love to start developing some mods for it. However, I've noticed from looking through the source that the current mods are either patches loaded from a JSON file (not very powerful) or are compiled in (not very extensible).
> 
> I'm wondering if you would be interested in a system where external mods can be loaded from DLLs rather than needing to be compiled in. This would allow people to develop mods without needing to have them merged into upstream, which could be better for some more specialist or larger mods, as well as reducing the amount of recompilation required during development.
> 
> Looking at `Mods.cpp`, this doesn't seem like it would be a particularly significant change. I would be happy to submit a pull request that adds code to automatically load each DLL in a directory at the end of `Mods::loadMods`.
> 
> Let me know if this is wanted and I'll get hacking! ^_^

It's a really cool idea, totally support this

**Comment by cursey** [comment]
How are you planning on exposing kanan's API to the plugins? You have to be careful with sending data across dll boundaries. It's probably enough to just dllexport everything in kanan core, dllexport the important bits in kanan, and see if theirs an option to dllexport imgui. Then each plugin could just link against kanan.lib and maybe it'd all work as long as everything was compiled with the same compiler version.

**Comment by pinkestt** [comment]
Yes, that will likely be the biggest challenge. I'll need to refresh my memory on C++'s calling conventions, as well as read up on how Windows handles DLLs since I'm more used to Linux.

Hopefully, it'd simply be a matter of linking Kanan into the plugins as a library, but as you say, the compiler will probably play a big role in whether that works or not.

**Comment by pinkestt** [comment]
From a quick look at StackOverflow, it looks like the biggest problem will be the fact that C++ doesn't specify the structure of classes. One solution to this could be to expose Kanan's API through a C interface, which is more standardized.

To make the API usage from plugins more similar to the internal API, some header-only wrapper classes could be created on the other end.

**Comment by cursey** [comment]
I would rather not expose a C API for kanan if at all possible. You can export C++ class methods, and like I said as long as the compiler matches it should work without much fuss (as long as care is taken passing data across dll boundaries). I don't particularly care about supporting other compilers. On Windows it's a decent expectation that a developer will be able to use MSVC and kanan only comes with build files for Visual Studio.

**Comment by pinkestt** [comment]
Sounds reasonable. Personally I use MinGW as I'm more familiar with GCC, but like you say, provided the same compiler is used to compile both parts it shouldn't matter.

**Comment by cursey** [comment]
Well, I just think it'd be a lot less work than creating a C API.
